xcdat/dataset.py

```diff
@@ -500,142 +413,102 @@ def _postprocess_dataset(
 
     Parameters
     ----------
-    arr : np.ndarray
-        An array of time axis values (e.g., coords, bounds)
+    da : xr.DataArray
+        An time-based DataArray (e.g,. coordinates, bounds)
 
     Returns
     -------
     bool
     """
-    if time_arr.ndim == 1:
-        vals = time_arr[0]
-    else:
-        vals = time_arr.flat[0]
+    units = da.encoding.get("units")
+    calendar = da.encoding.get("calendar")
 
-    is_decoded = time_arr.dtype == np.dtype("datetime64[ns]") or isinstance(
-        vals, cftime.datetime
-    )
+    return calendar is not None and units is not None
 
-    return is_decoded
 
+def _decode_time(da: xr.DataArray) -> xr.Variable:
+    """Lazily decodes a DataArray of numerically encoded time values with cftime.
 
-def _split_time_units_attr(units_attr: str) -> Tuple[str, str]:
-    """Splits the time coordinates' units attr into units and reference date.
+    The ``xr.DataArray` is` converted to an ``xr.Variable`` so that
+    ``xr.coding.variables.lazy_elemwise_func`` can be leveraged to lazily decode
+    time.
+
+    This function is based on ``xarray.coding.times.CFDatetimeCoder.decode``.
 
     Parameters
     ----------
-    units_attr : str
-        The units attribute (e.g., "months since 1800-01-01").
+    coords : xr.DataArray
+        A DataArray of numerically encoded time values.
 
     Returns
     -------
-    Tuple[str, str]
-        The units (e.g, "months") and the reference date (e.g., "1800-01-01").
+    xr.Variable
+        A Variable of time values decoded as ``cftime`` objects.
+    """
+    variable = xr.core.variable.as_variable(da)
+    dims, data, attrs, encoding = xr.coding.variables.unpack_for_decoding(variable)
 
-    Raises
-    ------
-    KeyError
-        If the time units attribute was not found.
+    units = xr.coding.variables.pop_to(attrs, encoding, "units")
+    calendar = xr.coding.variables.pop_to(attrs, encoding, "calendar")
 
-    ValueError
-        If the time units attribute is not of the form `X since Y`.
-    """
-    if "since" in units_attr:
-        units, reference_date = units_attr.split(" since ")
-    else:
-        raise ValueError(
-            "This dataset does not have time coordinates of the form 'X since Y'."
-        )
+    transform = partial(_get_cftime_coords, units=units, calendar=calendar)
+    data = xr.coding.variables.lazy_elemwise_func(data, transform, np.dtype("object"))
 
-    return units, reference_date
+    return xr.Variable(dims, data, attrs, encoding)
 
 
-def _get_cftime_coords(
-    ref_date: str, offsets: np.ndarray, calendar: str, units: str
-) -> np.ndarray:
+def _get_cftime_coords(offsets: np.ndarray, units: str, calendar: str) -> np.ndarray:
     """Get an array of `cftime` coordinates starting from a reference date.
 
+    This function supports multidimensional arrays such as bounds by flattening
+    them first, decoding, then reshaping back to their original shapes.
```

Comment:
This was inspired by xarray and I thought it was genius.

https://github.com/pydata/xarray/blob/076bd8e15f04878d7b97100fb29177697018138f/xarray/coding/times.py#L269